### PR TITLE
refactor helper to use in cli in CLD

### PR DESCRIPTION
**Pull request by krehermann**
[DPA-1396](https://smartcontract-it.atlassian.net/browse/DPA-1396)

refactor to support timelock runner as cli
### Supports

https://github.com/smartcontractkit/chainlink-deployments/pull/333

[DPA-1396]: https://smartcontract-it.atlassian.net/browse/DPA-1396?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: deployment/common/changeset/mcms_helpers.go [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+func RunTimelockExecutor(env deployment.Environment, executor *mcms.Executor, timelockContracts *TimelockExecutionContracts, sel uint64) error {
+	// TODO: This sort of helper probably should move to the MCMS lib.
+	// Execute all the transactions in the proposal which are for this chain.
+	if len(executor.Operations[mcms.ChainIdentifier(sel)]) == 0 {
+		return fmt.Errorf("no operations for chain %d", sel)
+	}
+	for _, chainOp := range executor.Operations[mcms.ChainIdentifier(sel)] {
+		for idx, op := range executor.ChainAgnosticOps {
+			if bytes.Equal(op.Data, chainOp.Data) && op.To == chainOp.To {
+				opTx, err2 := executor.ExecuteOnChain(env.Chains[sel].Client, env.Chains[sel].DeployerKey, idx)
```
Comment (krehermann) [inline deployment/common/changeset/mcms_helpers.go]: @akhilchainani  cli impl failing here

File: deployment/common/changeset/mcms_helpers.go [hunk 2]
```diff
@@ -0,0 +1,177 @@
+package changeset
```
Comment (connorwstein) [inline deployment/common/changeset/mcms_helpers.go]: maybe belongs in the propose dir? Or can kill that dir and move it into mcms_helpers.go

File: deployment/common/changeset/mcms_helpers.go [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+)
+
+// TimelockExecutionContracts is a helper struct for executing timelock proposals. it contains
+// the timelock and call proxy contracts.
+type TimelockExecutionContracts struct {
+	Timelock  *owner_helpers.RBACTimelock
+	CallProxy *owner_helpers.CallProxy
+}
+
+// NewTimelockExecutionContracts creates a new TimelockExecutionContracts struct.
```
Comment (connorwstein) [inline deployment/common/changeset/mcms_helpers.go]: I think you can use MaybeLoadMCMSWithTimelockState

File: deployment/common/changeset/mcms_helpers.go [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+	ChainSelector     uint64
+	// BlockStart is optional. It filter the timelock scheduled events.
+	// If not provided, the executor assumes that the operations have not been executed yet
+	// executes all the operations for the given chain.
+	BlockStart *uint64
+	BlockEnd   *uint64
+}
+
+func (cfg RunTimelockExecutorConfig) Validate() error {
+	if cfg.Executor == nil {
```
Comment (connorwstein) [inline deployment/common/changeset/mcms_helpers.go]: MaybeLoadMCMSWithTimelockState.Validate

File: deployment/common/changeset/test_helpers.go [hunk 5]
```diff
@@ -91,6 +91,7 @@ func ApplyChangesets(t *testing.T, e deployment.Environment, timelockContractsPe
 			NodeIDs:           e.NodeIDs,
 			Offchain:          e.Offchain,
 			OCRSecrets:        e.OCRSecrets,
+			GetContext:        e.GetContext,
```
Comment (cgruber) [inline deployment/common/changeset/test_helpers.go:95]: Dammit, another one. Shouldn't we be using NewEnvironment here instead of priming the struct anyway? Sigh.

(Not a review, just drive-by frustration)

File: deployment/common/changeset/state.go [hunk 6]
```diff
@@ -19,32 +19,18 @@ import (
...
-	BypasserMcm  *owner_helpers.ManyChainMultiSig
-	ProposerMcm  *owner_helpers.ManyChainMultiSig
-	Timelock     *owner_helpers.RBACTimelock
-	CallProxy    *owner_helpers.CallProxy
+	*proposalutils.MCMSWithTimelockContracts
 }
 
-// Validate checks that all fields are non-nil, ensuring it's ready
-// for use generating views or interactions.
-func (state MCMSWithTimelockState) Validate() error {
```
Comment (krehermann) [inline deployment/common/changeset/state.go:31]: mv with contracts to proposalutils

File: deployment/common/proposalutils/mcms_helpers.go [hunk 7]
```diff
@@ -0,0 +1,273 @@
...
+		panic(err)
+	}
+	lggr.Debug("scheduled", "event", string(b))
+}
+
+// MCMSWithTimelockContracts holds the Go bindings
+// for a MCMSWithTimelock contract deployment.
+// It is public for use in product specific packages.
+// Either all fields are nil or all fields are non-nil.
+type MCMSWithTimelockContracts struct {
```
Comment (krehermann) [inline deployment/common/proposalutils/mcms_helpers.go:184]: to use MaybeLoad.. and avoid import cycle need to defn this in this package. mv of `MCMSWithTimelockState`

File: deployment/ccip/changeset/accept_ownership_test.go [hunk 8]
```diff
@@ -9,9 +9,11 @@ import (
 	"golang.org/x/exp/maps"
 
 	commonchangeset "github.com/smartcontractkit/chainlink/deployment/common/changeset"
+	"github.com/smartcontractkit/chainlink/deployment/common/proposalutils"
 )
 
 func Test_NewAcceptOwnershipChangeset(t *testing.T) {
+	t.Parallel()
```
Comment (krehermann) [inline deployment/ccip/changeset/accept_ownership_test.go:16]: ci is timing out. parallelize many ccip tests

File: deployment/common/changeset/mcms_helpers.go [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+		return fmt.Errorf("no operations for chain %d", cfg.ChainSelector)
+	}
+	return nil
+}
+
+// RunTimelockExecutor runs the scheduled operations for the given chain.
+// If the block start is not provided, it assumes that the operations have not been scheduled yet
+// and executes all the operations for the given chain.
+// It is an error if there are no operations for the given chain.
+func RunTimelockExecutor(env deployment.Environment, cfg RunTimelockExecutorConfig) error {
```
Comment (cgruber) [inline deployment/common/changeset/mcms_helpers.go]: Sigh. It makes me nervous having this all accessible to changesets, as I want to avoid changesets themselves from ever executing this logic. But I guess this is fine for now - I agree, it probably ought to go into MCMS library though. We can move it there later.

**Review by cgruber — DISMISSED** [review]

File: deployment/common/changeset/mcms_helpers.go (hunk 9 above)
Comment (krehermann) [reply deployment/common/changeset/mcms_helpers.go]: do either of these make it better:
- in a real env (with non zero timelock) it doesn't matter b/c the duration must pass while in the timelock
- moving pkgs out of changeset

File: deployment/ccip/changeset/test_assertions.go [hunk 10]
```diff
@@ -221,8 +221,8 @@ func ConfirmCommitForAllWithExpectedSeqNums(
 			return false
 		}
 	},
-		3*time.Minute,
-		1*time.Second,
+		tests.WaitTimeout(t),
```
Comment (krehermann) [inline deployment/ccip/changeset/test_assertions.go:224]: this fixed a lot of flakes

Comment (krehermann) [reply deployment/ccip/changeset/test_assertions.go:224]: it may also be related to the recent ccip smoke test nightmares

File: deployment/common/changeset/internal/mcms_test.go [hunk 11]
```diff
@@ -36,9 +37,9 @@ func TestDeployMCMSWithTimelockContracts(t *testing.T) {
 	_, err := internal.DeployMCMSWithTimelockContracts(lggr,
 		chains[chainsel.TEST_90000001.Selector],
 		ab, types.MCMSWithTimelockConfig{
-			Canceller:        changeset.SingleGroupMCMS(t),
-			Bypasser:         changeset.SingleGroupMCMS(t),
-			Proposer:         changeset.SingleGroupMCMS(t),
+			Canceller:        proposalutils.SingleGroupMCMS(t),
+			Bypasser:         proposalutils.SingleGroupMCMS(t),
+			Proposer:         proposalutils.SingleGroupMCMS(t),
 			TimelockMinDelay: big.NewInt(0),
```
Comment (graham-chainlink) [inline deployment/common/changeset/internal/mcms_test.go]: Why not just `proposalutils.SingleGroupTimelockConfig(t)` ?

**Review by jmank88 — DISMISSED** [review]

File: deployment/common/changeset/internal/mcms_test.go (hunk 11 above)
Comment (krehermann) [reply deployment/common/changeset/internal/mcms_test.go]: oversight. thanks

**Review by graham-chainlink — APPROVED** [review]

**Review by connorwstein — APPROVED** [review]

File: deployment/common/changeset/mcms_helpers.go (hunk 9 above)
Comment (cgruber) [reply deployment/common/changeset/mcms_helpers.go]: I think both help, yeah.
